workspace/resource_repo.go

```diff
@@ -0,0 +1,203 @@
+	HeadCommitID string `json:"head_commit_id,omitempty" tf:"computed,alias:commit_hash"`
+}
+
+// ID returns job id as string
+func (r ReposInformation) RepoID() string {
+	return fmt.Sprintf("%d", r.ID)
+}
+
+type createRequest struct {
+	Url      string `json:"url"`
+	Provider string `json:"provider"`
+	Path     string `json:"path,omitempty"`
+}
+
+func (a ReposAPI) Create(r createRequest) (ReposInformation, error) {
+	var resp ReposInformation
+	if r.Provider == "" { // trying to infer Git Provider from the URL
+		r.Provider = getProviderFromUrl(r.Url)
+	}
+	if r.Provider == "" {
+		return resp, fmt.Errorf("git_provider isn't specified and we can't detect provider from URL")
+	}
+
+	err := a.client.Post(a.context, "/repos", r, &resp)
+	return resp, err
+}
+
+func (a ReposAPI) Delete(id string) error {
+	return a.client.Delete(a.context, fmt.Sprintf("/repos/%s", id), nil)
+}
+
+func (a ReposAPI) Update(id string, r map[string]string) error {
+	if len(r) == 0 {
+		return nil
+	}
+	// TODO: update may change ONE OF (url AND provider (optional)), (path), or (branch OR tag).
+	// for URL/provider force re-create as there are limits on what could be done for changing URL/provider
+	if path, ok := r["path"]; ok {
+		err := a.client.Patch(a.context, fmt.Sprintf("/repos/%s", id), map[string]string{"path": path})
+		if err != nil {
+			return err
+		}
+		delete(r, "path")
+	}
+	return a.client.Patch(a.context, fmt.Sprintf("/repos/%s", id), r)
+}
+
+func (a ReposAPI) Read(id string) (ReposInformation, error) {
+	var resp ReposInformation
+	err := a.client.Get(a.context, fmt.Sprintf("/repos/%s", id), nil, &resp)
+	return resp, err
+}
+
+type ReposListResponse struct {
+	NextPageToken string             `json:"next_page_token,omitempty"`
+	Repos         []ReposInformation `json:"repos"`
+}
+
+func (a ReposAPI) List(prefix string) ([]ReposInformation, error) {
+	req := map[string]string{}
+	if prefix != "" {
+		req["path_prefix"] = prefix
+	}
+	reposList := []ReposInformation{}
+	for {
+		var resp ReposListResponse
+		err := a.client.Get(a.context, "/repos", req, &resp)
+		if err != nil {
+			return nil, err
+		}
+		reposList = append(reposList, resp.Repos...)
+		if resp.NextPageToken == "" {
+			break
+		}
+		req["next_page_token"] = resp.NextPageToken
+	}
+	return reposList, nil
+}
+
+func (a ReposAPI) ListAll() ([]ReposInformation, error) {
+	return a.List("")
+}
+
+var gitProvidersMap = map[string]string{
+	"github.com":    "gitHub",
+	"dev.azure.com": "azureDevOpsServices",
+	"gitlab.com":    "gitLab",
+	"bitbucket.org": "bitbucketCloud",
+}
+
+func getProviderFromUrl(uri string) string {
+	provider := ""
+	u, err := url.Parse(uri)
+	if err == nil {
+		provider = gitProvidersMap[strings.ToLower(u.Host)]
+	}
+	return provider
+}
+
+func ResourceRepo() *schema.Resource {
+	s := common.StructToSchema(ReposInformation{}, func(s map[string]*schema.Schema) map[string]*schema.Schema {
+		s["url"].ForceNew = true
```

Comment:
i wonder if we should add `force_new` to reflect_resource

Reply:
I thought about that, just didn't want to pollute the PR for specific functionality